ls1intum/Artemis
# UI change of minimized programming instructions header to look more like problem statement header

### Checklist
- [x] We tested *all* changes and *all* related features with different users (student, tutor, editor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Client: We followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] Client: We added multiple screenshots/screencasts of my UI changes
- [x] Client: We translated all newly inserted strings into English and German.

### Motivation and Context

The first couple of images show the current UI for instructions (i.e. problem statement for programming exercises).
If you click on the header you can minimize the instructions to give the code editor more space.

programming : normal view [old]            |  programming : collapsed view [old]
:-------------------------:|:-------------------------:
![](https://user-images.githubusercontent.com/65814168/125746600-e648c8be-1b53-487b-8185-40cb8b3f59a8.png)  | ![](https://user-images.githubusercontent.com/65814168/125746785-c4c771d0-5fa5-4139-b572-ffc89fb4d2fb.png)

If you compare this UI with the other exercise types you can spot that the design is quite different therefore it breaks the homogeneous look and feel of the exam mode.

other exercises : normal view  | other exercises : collapsed view
:-------------------------:|:-------------------------:
![](https://user-images.githubusercontent.com/65814168/126064363-f0d5471d-12b2-48c9-b4e3-59adde4d0355.png) | ![](https://user-images.githubusercontent.com/65814168/126064420-8b228ee6-751c-4651-a5f3-d41d7fc13abb.png)

This issue has been spotted and requested to be fixed during the developer meeting's exam testing session (13.07.2021).

### Description

We changed the instructions header to resemble the other exercise type's problem statement as close as possible.
<details>
<summary> remarks about the code structure </summary>
<p> --------------------------------------------------- </p>
<p>At first, we looked at the implementation of the problem statement header for text exercises and tried to reuse it for programming exercises. </p>
<p>The problem statement itself is located in `text-exam-submission.component.html` which itself is embedded in `resizeable-container.component.html` (containers are connected to CSS grids that determine the location of elements).</p>
<p>The logic behind the collapsed/minimized and open/normal view is based on a simple `*ngIf="collapsed; then collapsedBody; else expandedBody` rule. I.e. a private component property `collapsed` determines what kind of element should be shown/created. </p>
This reasoning is quite easy, and we thought that a similar logic will be used for programming exercises.

Programming exercises' instructions are located inside `programming-exam-submission.component.html` (right next to the text exercise's one when you look at the file hierarchy, i.e. inside **../exam/participate/exercises**)
But the logic is sadly way more complex, coupled, and confusing to understand. 
Multiple different TS, CSS, HTML files that are all interconnected => high coupling
It is quite hard to understand what influences what and how.
Instead of a simple *ngIf case the collapsing is done via event handling and grid methods that resize the visible element.
<p> => this code was mostly written in 2019 and really needs refactoring </p>

<p> --------------------------------------------------- </p>
</details>
We changed the header text from "Instructions" to "Problem Statement" (including German translation).
_The other (visual) differences can be seen more easily than described by words._

### Result
programming : normal view [new]            |  programming : collapsed view [new]
:-------------------------:|:-------------------------:
![](https://user-images.githubusercontent.com/65814168/126065843-b8f3b229-3914-45bf-b386-2093f112bc01.png)  | ![](https://user-images.githubusercontent.com/65814168/126065873-59f36e68-20d6-41da-ae3c-f32937673d8e.png)

The new problem statement for programming exercises still has two minor differences from the other exercise types:
- the font size is a bit larger for programming exercises, we decided to keep it because otherwise, it will not fit with the other texts on the online code editor.
- there is a visible icon "||" that is used for changing the size of the opened window when dragged. We decided to keep this because it also stays when the other collapsable elements are minimized (e.g. file browser). 

### Steps for Testing


1. Log in to Artemis
2.a Go to an exam with a programming exercise (create one if you can not find one)
2.b Go to a normal course programming exercise and click on online code editor (it uses the same UI)
3. Look at the instructions (header) : it has to look like the old one
4. Click on it and make sure it minimizes correctly (and opens back up successfully as well)
5. The minimized instructions should look more in tune with the other exercise types (problem statement header)
6. Make sure the final result looks like our screenshots.
